### DISCUSSION TITLE
Should IDE created classes be synched back to project?

### DISCUSSION THREAD

**Discussion by paulrobinson** [jbangdev/jbang#1087] [Q&A] [answered]
When I edit my JBang enabled Java file with `jbang edit --open=idea run.java` and then add some extra classes into the `src` directory (in the IntelliJ IDEA project tree): Should they be appearing in my source project somewhere?

My observation is that these additional, IDE created, source files are only available when I execute the run.java from my IDE. Is this expected behaviour?

**Comment by quintesse** [comment, marked as answer]
It's expected in the way that we know about it, but it's not exactly what we would like :-)
We're currently trying to define how we want `edit` to behave and how to implement that.

Just to give a bit of an idea as to the kind of problems we're facing here:

- JBang allows files without `.java` extension (eg `jbang run  myprog`)
- JBang allows files to not be in their proper package folder (eg. a file defining `package my.org` does not need to be in  a folder called `my/org`)
- JBang allows completely unrelated (and possibly conflicting) files to reside in the same folder (eg. several files define a non-public class with the same name)

These are all issues that would trip up an IDE, so to make sure that everything works okay we create a separate project directory with the proper folder structure and with proper file names (which are links back to the original files).

This works fine (as long as you're on Linux or Mac that is, on Windows is a whole other story) as long as you don't add any new files (as you noticed) and as long as you're happy to just be able to edit a single file (and it's defined //SOURCES dependencies) as opposed to be able to edit all files in a folder.

So we know about the problem, we're just not sure yet about the solution :-)

**Reply by paulrobinson** [reply]
Thanks @quintesse for the detailed response. I can see why this is a complex issue. It's good to know it's not a bug and that I'm not doing something wrong. 

In my case, the extra classes are just some simple POJO data structures. So I've just added them as inner classes for now.
